### DISCUSSION TITLE
建议给自动分表加上手动取消聚合的功能

### DISCUSSION THREAD

**Discussion by woodcoal** [dotnetcore/FreeSql#1335] [Ideas]
当设置了自动分表后，默认 select 会自动聚合所有分表，分页就是一个很大的问题。
同时有时候我明知某些表不会存在数据，或者说我只想查询某个表的时候，就会造成麻烦。

比如：
```
[Table(Name = "as_table_log_{yyyyMM}", AsTable = "createtime=2022-1-1(1 month)")]
class AsTableLog{}
```

会自动生成：
```
as_table_log_202201
as_table_log_202202
as_table_log_202203
...
```

如果我使用 fsql.Select<AsTableLog> 则会默认将所有表聚合查询，当我只想查询 as_table_log_202202 这个表的时候就没有办法处理了。

建议在查询的时候增加一个参数来限定时间区域，比如说 
1.  withRange(2022-03-01,2022-04-30) 就只聚合 as_table_log_202203 与 as_table_log_202204 两个表
2.  withDate(2022-03-15) 就只查询 as_table_log_202203 这个表

因为自动分表是按时间做模板处理的，所以带上时间参数就很容易匹配对应表的名称

**Comment by woodcoal** [comment]
目前使一个简单粗暴的办法 fsql.Select<>.AsTable
使用了 AsTable 原来查询中默认的所有关联都自动取消了，查询以 AsTable 中设置为准

**Comment by 2881099** [comment]
fsql.Select\<T\>().Where(a=> a.time.Between(2022-03-01,2022-04-30)) 就只聚合 as_table_log_202203 与 as_table_log_202204 两个表

withDate(2022-03-15) 就只查询 as_table_log_202203 这个表，至于这个可以AsTable 也可以同上处理，或者定义自定义方法解析 .Where(a=> SqlFunc.SameDate(a.Time, date))

**Reply by woodcoal** [reply]
> fsql.Select<T>().Where(a=> a.time.Between(2022-03-01,2022-04-30)) 就只聚合 as_table_log_202203 与 as_table_log_202204 两个表
> 
> withDate(2022-03-15) 就只查询 as_table_log_202203 这个表，至于这个可以AsTable 也可以同上处理，或者定义自定义方法解析 .Where(a=> SqlFunc.SameDate(a.Time, date))

测试了下，a.time 区间来聚合对应表是正确的，但是貌似也不是很准确，比如说，我见了一个年为单位的分表

[Table(Name = "Message_{yyyy}", AsTable = "TimeAdd=2015-1-1(1 year)")]

当查询 TimeAdd 从 2020-3-1 到 2020-5-1 时，原则上应该值查询 Message_2020，但是实际 Message_2020，Message_2021，Message_2022 三个都聚合了，都是从最小时间起到当前日期止之间的表，而不是指定的起始时间

**Comment by 2881099** [comment]
目前会不会聚合是根据 sql where 条件做处理的

**Reply by woodcoal** [reply]
> 目前会不会聚合是根据 sql where 条件做处理的

清楚了，谢谢！
